pvlib/ivtools.py

```diff
@@ -0,0 +1,351 @@
+    In-Field Photovoltaic I-V Curves on a Single Board Computer, 46th IEEE
+    Photovoltaic Specialist Conference, Chicago, IL, 2019
+    """
+
+    # If not provided, extract v_oc, i_sc, v_mp and i_mp from the IV curve data
+    if v_oc is None:
+        v_oc = voltage[-1]
+    if i_sc is None:
+        i_sc = current[0]
+    if v_mp_i_mp is not None:
+        v_mp, i_mp = v_mp_i_mp
+    else:
+        v_mp, i_mp = _find_mp(voltage, current)
+
+    # Find beta0 and beta1 from linear portion of the IV curve
+    beta0, beta1 = _find_beta0_beta1(voltage, current, vlim, v_oc)
+
+    # Find beta3 and beta4 from the exponential portion of the IV curve
+    beta3, beta4 = _find_beta3_beta4(voltage, current, beta0, beta1, ilim,
+                                     i_sc)
+
+    # calculate single diode parameters from regression coefficients
+    return _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp,
+                                     v_oc)
+
+
+def _find_mp(voltage, current):
+    """
+    Finds voltage and current at maximum power point.
+
+    Parameters
+    ----------
+    voltage : ndarray
+        1D array containing voltage at each point on the IV curve, increasing
+        from 0 to v_oc inclusive, of `float` type [V]
+
+    current : ndarray
+        1D array containing current at each point on the IV curve, decreasing
+        from i_sc to 0 inclusive, of `float` type [A]
+
+    Returns
+    -------
+    v_mp, i_mp : tuple
+        voltage ``v_mp`` and current ``i_mp`` at the maximum power point [V],
+        [A]
+    """
+    p = voltage * current
+    idx = np.argmax(p)
+    return voltage[idx], current[idx]
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
+
+
+def _find_beta0_beta1(v, i, vlim, v_oc):
+    # Get intercept and slope of linear portion of IV curve.
+    # Start with V =< vlim * v_oc, extend by adding points until slope is
+    # negative (downward).
+    beta0 = np.nan
+    beta1 = np.nan
+    first_idx = np.searchsorted(v, vlim * v_oc)
+    for idx in range(first_idx, len(v)):
+        coef = np.polyfit(v[:idx], i[:idx], deg=1)
+        if coef[0] < 0:
+            # intercept term
+            beta0 = coef[1].item()
+            # sign change of slope to get positive parameter value
+            beta1 = -coef[0].item()
+            break
+    if any(np.isnan([beta0, beta1])):
+        raise RuntimeError("Parameter extraction failed: beta0={}, beta1={}"
+                           .format(beta0, beta1))
+    else:
+        return beta0, beta1
+
+
+def _find_beta3_beta4(voltage, current, beta0, beta1, ilim, i_sc):
+    # Subtract the IV curve from the linear fit.
+    y = beta0 - beta1 * voltage - current
+    x = np.array([np.ones_like(voltage), voltage, current]).T
+    # Select points where y > ilim * i_sc to regress log(y) onto x
+    idx = (y > ilim * i_sc)
+    result = np.linalg.lstsq(x[idx], np.log(y[idx]), rcond=None)
```

Comment:
Out of curiosity, why `lstsq` here and `polyfit` above?

Reply:
`polyfit` is single variable only. Here we need multiple regression.

Reply:
Oh yes, fairly obvious now that you mention it.